[ADD] 진료과목, 의사 리스트 기능 추가 완료(ADD: department, doctor list View)

jhwang0801 opened the pull request:
## :: 최근 작업 주제
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 
### - 진료 과목 리스트(DepartmentsListView) & 해당 진료과목 소속 의사 리스트(DoctorListView) 기능 구현

<br />

## :: 구현 사항 설명 (순서대로 나열)
### 1. 기존 파일 에러 수정
- 이전 파일들 중 에러가 나는 부분 수정
- `users/tests.py` 관련 부분 수정 완료 / 테스트 결과 모두 pass

### 2. Login_decorator 추가 완료
- 로그인 과정에서 발행된 토큰 값을 검증하는 `Decorator` 작성
- 예외처리(Invalid_token, invalid_user, expired_token)

### 3. 진료과목 리스트 뷰 (DepartmentsListView)
- `List Comprehension` 적용
- 진료 과목 이미지는 로컬서버에서 불러올 수 있도록 지정
    - IP 주소의 경우,`LOCAL_PATH` 환경변수 지정하여 감춰줌

### 4. 의사 리스트 뷰 (DoctorListView)
- `List Comprehension` 적용
- `Path Parameter` 적용
- `Paginator` 적용
    - 한 페이지 당 의사 6명만 보일 수 있도록 함
    - "Pagination" 관련된 예외처리 적용
        - `PageNotAnInteger` : 페이지의 값이 정수가 아닐 때 발생
        -  `EmptyPage` : 해당 페이지의 내용에 값이 없을 때 발생
- `ORM 최적화`(select_related() 적용
    - *<DB 안에 데이터가 2개 있는 경우>*
![스크린샷 2022-06-27 오전 11 28 33](https://user-images.githubusercontent.com/75832544/175863649-40049c8e-11b3-4b97-884e-f74260101265.png)
    - *<DB 안에 데이터가 3개로 증가시킨 경우>* (데이터가 늘어날 때마다 Query 횟수가 '3' 증가함)
![스크린샷 2022-06-27 오전 11 29 34](https://user-images.githubusercontent.com/75832544/175863749-1e2bb927-230c-4363-a29f-72738ec524d8.png)
    - *<select_related() method를 적용하여 ORM 최적화를 진행한 경우>
![스크린샷 2022-06-27 오전 11 28 49](https://user-images.githubusercontent.com/75832544/175863920-abc8f16a-9e0c-4bff-9c44-c77f3e815c20.png)

### 5. 테스트 코드 작성
![스크린샷 2022-06-27 오전 11 36 25](https://user-images.githubusercontent.com/75832544/175864087-94c1b435-a062-441d-930e-e8fcaccd444a.png)


<br />

## :: Postman 결과
#### (1) 진료 과목 리스트 성공
![image](https://user-images.githubusercontent.com/75832544/175864363-102fc665-85b5-4bed-bad9-8bc83d81bacf.png)

#### (2) 의사 리스트 성공
![image](https://user-images.githubusercontent.com/75832544/175864470-f4ad047a-21ac-4351-a800-4f5ae67b7533.png)

#### (3) 의사 리스트 "EmptyPage Error"
![image](https://user-images.githubusercontent.com/75832544/175864552-98830223-50d3-42da-a136-d3adf6b6c038.png)

#### (4) 의사 리스트 "PageNotAnInteger Error"
![image](https://user-images.githubusercontent.com/75832544/175864595-4ef601e5-4698-42c0-a9cd-0cbab4d86486.png)


<br />

## :: 질문
### 1. `.values()` Method 사용 방법
- 첫날 알려주시기로 `.all()` 이나 `.filter()`를 사용하는 경우, `.values()`와 `.annotate()`로 작성이 가능하다고 말씀해주셔서 방법을 적용해보려 했으나, 진료과목 리스트 뷰의 경우에는 "thumbnail"(이미지 경로를 알려주어야함), 의사리스트 뷰의 경우에는 "id"를 제외한 모든 값을 어떻게 처리해 주어야 제가 원하는 딕셔너리 형태로 뽑을 수 있을 지 모르겠습니다.
제가 작성한 지금 이 코드에도 말씀해주셨던 부분을 적용하여 코드를 작성할 수 있는 지 조언을 주시면 감사하겠습니다.

jhwang0801 committed:
ADD: department, doctor list View

- users/tests.py 에러나는 부분 모두 수정 완료
- users/utils.py login_decorator 추가
- appointments/views.py DeparmentsListView 추가 완료(List Comprehension 적용)
- appointments/views.py DoctorListView 추가 완료(List Comprehension, ORM최적화, paginator 적용)
- appointments/tests.py 테스트코드 추가 완료

jasonkang14 commented on appointments/tests.py:
```diff
@@ -1,3 +1,114 @@
+                    {
+                        "id"       : 1,
+                        "name"     : "가정의학과",
+                        "thumbnail": "127.0.0.1:8000/media/department_thumbnail/family_medicine.png"
```
`http` 나 `https` 추가 안해줘도 잘 작동하나요? 
궁금해서 드리는 질문입니다~ 
브라우저에서는 자동으로 프로토콜을 입력해주니까 그냥 될 것 같기도 하네요

jhwang0801 replied:
`http`의 경우, 입력 없이도 브라우저에서는 잘 작동되는 걸 확인했습니다.

`https`는 SSL_PROTOCOL_ERROR로 인하여 확인은 하지 못했습니다.

jasonkang14 commented on appointments/views.py:
```diff
@@ -1,3 +1,43 @@
+        departments_list = [{
+            'id'       : department.id,
+            'name'     : department.name,
+            'thumbnail': settings.LOCAL_PATH + "department_thumbnail/" + f"{department.thumbnail}"
```
말씀하신 것처럼 여기서 annotate를 사용할 수 있습니다
django database funtion중에 [Concat](https://docs.djangoproject.com/en/4.0/ref/models/database-functions/#concat)을 사용하면 `thumnail` 값을 만들기 위해 사용하신 연산을 [annotate](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#annotate)를 사용할 수 있습니다. 

그리고 string `+`를 사용하는 concat보다 `f-string`을 사용하는 것이 더 효율적입니다.
f-string을 잘 사용하셨는데 string 전체를 감쌀 수 있습니다.
```python
'thumbnail' : f'{settings.LOCAL_PATH}department_thumbnail/{department.thumbnail}'
```
가독성을 생각하면 `settings.LOCAL_PATH`와 `department_thumbnail` 사이에 `/`가 들어가면 좋겠네요
하지만 이건 annotate 사용하시면 해결될 문제이니 고민하지 않으셔도 될 것 같습니다.

f-string을 사용하는 이유는
일단 속도가 더 빠릅니다 [stackoverflow](https://stackoverflow.com/a/38722192)
파이썬에서 string은 immutable하기 때문에 concat을 위해 string을 작성하면, 파이썬에서 각 스트링마다 메모리를 할당하기 때문입니다  [stackoverflow](https://stackoverflow.com/a/21640094)
이게 f-string이 어 빠른 이유가 될 수도 있겠네요

jhwang0801 replied:
DeparmentsListView, DoctorListView 모두 Concat, annotate를 사용해서 작성해 보았습니다.
코드 작성 이후 테스트 과정에서 문의드릴 사항이 생겼습니다.

아래와 같이 DoctorListView를 작성해보았는데요,
```python
page    = request.GET.get('page', 1)
doctors = Doctor.objects.select_related('user', 'department', 'hospital').filter(department_id=department_id)\
    .annotate(
        names        = Concat(V(''), 'user__name', output_field=CharField()),
        departments  = Concat(V(''), 'department__name', output_field=CharField()),
        hospitals    = Concat(V(''), 'hospital__name', output_field=CharField()),
        profile_imgs = Concat(V(f'{settings.LOCAL_PATH}/doctor_profile_img/'), 'profile_img', output_field=CharField())
    ).values('id', 'names', 'departments', 'hospitals', 'profile_imgs')

doctors_paginator = Paginator(doctors, 6).page(page).object_list

return JsonResponse({"result" : list(doctors_paginator)}, status=200)

```

작성 이후 제가 이전에 작성한 List Comprehension과 `execution time`이 얼마나 차이가 날까 궁금해서 돌려보았고, 아래와 같은 결과를 얻었습니다.
#### (1) List Comprehension execution time
![스크린샷 2022-06-27 오후 3 51 57](https://user-images.githubusercontent.com/75832544/175878935-99064b0a-63e6-482c-bce2-6a887765a0d5.png)

#### (2) Concat & annotate execution time
![스크린샷 2022-06-27 오후 3 52 35](https://user-images.githubusercontent.com/75832544/175878992-87d72b6a-d816-439a-bb68-3ec908367eb4.png)

제가 예상했던 것보다 차이가 거의 안났고, 오히려 List Comprehension 쪽의 평균 `execution time`이 낮다고 보여지는데, 혹시 이 부분이 DB에 데이터가 얼마 없기 때문에 그런것일까요?

아니면 제가 작성한 코드에 오류가 있기 때문일까요?

jasonkang14 replied:
execution time을 어떻게 측정하셨나요? 
지금 판단으로는 
컬럼의 숫자가 적기 때문에 `SELECT *`를 하는것과 `SELECT id, name` 등등 하는 것이 별 차이가 없어서 그런 것 같습니다.

jhwang0801 replied:
정확한 방법은 아니지만, 대략적인 확인을 하고싶어서

```python
import time
start = time.time()
end = time.time()
```

를 사용해서 측정해봤습니다.

jasonkang14 replied:
시간만 따지면 그 때 그 때 확인해서 list comprehension을 할지 아니면 annotate를 사용할지를 구분해야 할 것 같은데 
혹시 execution말고 위에 스크린샷 첨부하신 것처럼query시간을 한 번 봐주실 수 있을까요?

jhwang0801 replied:
`DoctorListView` 로 진행하였습니다.
#### (1) List Comprehension
![스크린샷 2022-06-27 오후 5 09 36](https://user-images.githubusercontent.com/75832544/175892200-322ae47d-9735-4d0c-95be-70d673db63b6.png)

#### (2) Concat & Annotate
![스크린샷 2022-06-27 오후 5 10 34](https://user-images.githubusercontent.com/75832544/175892253-f4da95c8-8323-4f05-9fff-863913e1c543.png)
위 스크린샷의 `UnorderedObjectListWarning`은 `.order_by('id')` 등 `order_by`만 정의해주면 해결되었고, 
결과는 아래와 같이 동일했습니다.
![image](https://user-images.githubusercontent.com/75832544/175892826-d9f2bdd8-a3af-4b1f-bea5-0e7a7c2daa31.png)


테스트에 사용한 query_decorator는 아래와 같습니다.
```python
def query_debugger(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        reset_queries()
        number_of_start_queries = len(connection.queries)
        start  = time.perf_counter()
        result = func(*args, **kwargs)
        end    = time.perf_counter()
        number_of_end_queries = len(connection.queries) - 2 if len(connection.queries) else 0
        print(f"-------------------------------------------------------------------")
        print(f"Function : {func.__name__}")
        print(f"Number of Queries : {number_of_end_queries-number_of_start_queries}")
        print(f"Finished in : {(end - start):.2f}s")
        print(f"-------------------------------------------------------------------")
        return result
    return wrapper
```

jasonkang14 replied:
음 네네 그럼 annotate로 가는 편이 코드 통일성 측면에서 더 좋을 것 같습니다

jhwang0801 replied:
로깅 설정 후 추가 정보 전달드리면,

#### (1) List Comprehension
- 총 디버그 횟수 : 3회
![스크린샷 2022-06-27 오후 5 30 19](https://user-images.githubusercontent.com/75832544/175895902-2b791d5b-f147-40c9-ad7a-e4cec96c993f.png)

#### (2) Concat & annotate
- 총 디버그 횟수 : 4회
![스크린샷 2022-06-27 오후 5 27 53](https://user-images.githubusercontent.com/75832544/175895947-e1523277-3a1c-4ffc-9934-7ee03d43627b.png)

결과가 위와 같습니다.

jhwang0801 replied:
`connection.queries` 결과 값 전달드립니다~

#### (1) List Comprehension
![스크린샷 2022-06-27 오후 5 40 18](https://user-images.githubusercontent.com/75832544/175897859-fefb1a39-b1bf-4364-821f-df740454d922.png)


#### (2) annotate
![스크린샷 2022-06-27 오후 5 41 04](https://user-images.githubusercontent.com/75832544/175897880-37b32075-732b-4193-a460-994fba9e0a6f.png)

입니다!

jasonkang14 commented on appointments/views.py:
```diff
@@ -1,3 +1,43 @@
+            'thumbnail': settings.LOCAL_PATH + "department_thumbnail/" + f"{department.thumbnail}"
+        } for department in Department.objects.all()]
+
+        return JsonResponse({"result" : departments_list, "user_name" : request.user.name}, status = 200)
```
`user_name`을 리턴하는 이유는 무엇인가요?

jhwang0801 replied:
![image](https://user-images.githubusercontent.com/75832544/175869223-e4c2531c-07f6-47a5-a203-d9053521fccc.png)

위 피그마의 진료과목 리스트와 관련하여 프론트엔드 분들과 회의 중,
"유저의 이름"을 데이터로 함께 전달해주면 좋을 것 같다는 요청이 있어 사용자 이름이 리턴될 수 있도록 하였습니다.

jasonkang14 replied:
REST의 규칙을 생각해보면, 하나의 view에서는 하나의 역할만 하는 편이 좋습니다. 
클라이언트에서 로그인이 되었는지 안되었는지를 파악하는 절차가 필요할텐데, 
로그인이 되었는지 안 되었는지를 파악하는 view에서 user_name을 클라이언트로 넘겨주는 편이 
REST에 더 적합한 것 같습니다. 

해당 View는 로그인 여부를 파악하기 위해 무조건 호출해야하니 추가로 작업해야 되는 부분이 아니기도 하구요. 
프런트와 협의해서 수정 부탁 드립니다!

jhwang0801 replied:
피드백 감사합니다!
프론트분들과 협의하여 수정하겠습니다!

jhwang0801 replied:
로그인 단계에서 로그인 시도하는 유저의 `id` 와 `name` 을 토큰 값과 함께 전달해주는 방향으로 진행하기로 프론트분들과 협의하였습니다.

차후 프론트엔드와 통신할 때, 다시 맞춰보고 수정이 필요한 경우 다시 방법을 찾아보기로 결정하였습니다.

jasonkang14 commented on appointments/views.py:
```diff
@@ -1,3 +1,43 @@
+    def get(self, request, department_id):
+        try: 
+            page        = request.GET.get('page', 1)
+            doctors     = Doctor.objects.select_related('user', 'department', 'hospital').filter(department_id=department_id)
```
`selet_related` 👍

jasonkang14 commented on appointments/views.py:
```diff
@@ -1,3 +1,43 @@
+                'profile_img': settings.LOCAL_PATH + "doctor_profile_img/" + f"{doctor.profile_img}"
+            } for doctor in doctors]
+
+            doctors_paginator = Paginator(doctor_list, 6).page(page).object_list
```
👍

jasonkang14 commented on users/utils.py:
```diff
@@ -20,6 +21,24 @@ def check_duplicate_email(email):
+            user         = CustomUser.objects.get(id = payload['user_id'])
+            request.user = user
+
+        except jwt.exceptions.DecodeError: 
```
`INVALID_TOKEN`도 401이면 좋겠습니다

jhwang0801 replied:
아! `INVALID_TOKEN`도 인증되지 않은 상태에서 리소스를 요청하는 경우이니, "401"로 변경하도록 하겠습니다!

jasonkang14 reviewed (changes requested):
1. annotate에 관련된 부분은 comment를 확인해주세요!
2. **구현 사항 설명은 굳이 작성하지 않으셔도 됩니다**. 어떤 의도를 가지고 코드를 작성했는지는 제가 보고 파악할 수 있어야 합니다. 코드를 보고 이해를 못한다면 그건 제가 실력이 부족한거라서요~ 
다음 PR부터는 작성하지 마시고, 혹시 이해가 잘 안되는 부분이 있다면 제가 PR comment로 질문을 드리도록 하겠습니다. 

고생하셨습니다!

jhwang0801 commented:
> 1. annotate에 관련된 부분은 comment를 확인해주세요!
> 2. **구현 사항 설명은 굳이 작성하지 않으셔도 됩니다**. 어떤 의도를 가지고 코드를 작성했는지는 제가 보고 파악할 수 있어야 합니다. 코드를 보고 이해를 못한다면 그건 제가 실력이 부족한거라서요~
>    다음 PR부터는 작성하지 마시고, 혹시 이해가 잘 안되는 부분이 있다면 제가 PR comment로 질문을 드리도록 하겠습니다.
> 
> 고생하셨습니다!

jhwang0801 committed:
Modified(1): department, doctor list View

- List Comprehension -> Concat & annotate 로 변경
- user_name 리턴 값 제거, users/views.py 로그인 뷰에서 유저 id, name 데이터 리턴하도록 변경
- 변경된 뷰에 따라서 테스트코드 변경 완료 / unittest 완료
- Invalid_token 400 에러 -> 401에러 변경 완료

jasonkang14 commented on users/tests.py:
```diff
@@ -196,7 +200,9 @@ def test_success_web_doctor_login(self):
             'message'     : 'SUCCESS_LOGIN',
-            'access_token': generate_jwt(CustomUser.objects.get(is_doctor=True))
+            'access_token': generate_jwt(CustomUser.objects.get(is_doctor=True)),
+            'user_id'          : CustomUser.objects.get(is_doctor=True).id,
```
테스트코드라서 상관은 없지만 이런식으로 작성하면 DB에 두 번 접근하게돼서 비효율적입니다. 
custom user를 변수에 담고, 해당 변수에서 `.id`, `.name`을 하시는 편이 좋습니다.

jhwang0801 replied:
피드백 감사합니다! 변수 설정하도록 하겠습니다!

jasonkang14 reviewed (approved):
LGTM
